File: src/apkinfo.py
#!/usr/bin/env python
 # -*- coding: UTF-8 -*-
import os
import sys

default_encoding = 'utf-8'
if sys.getdefaultencoding() != default_encoding:
	reload(sys)
	sys.setdefaultencoding(default_encoding)

#打印对象
def prn_obj(obj):
    print ', '.join(['%s:%s' % item for item in obj.__dict__.items()])

#换行打印对象
def prn_obj2(obj):
    print '\n'.join(['%s\t%s' % item for item in obj.__dict__.items()])

#解析存在兼职对的字符串
def analyze_key_value(xml_str):
	key_dict = {}
	pacakge_array = xml_str.split(' ')
	for pacakge_item in pacakge_array:
		item_array = pacakge_item.split('=')
		if len(item_array)!=2 :
			continue
		key_dict[item_array[0]] = item_array[1].replace('\'','').replace('\n','')
	return key_dict

#apk的信息
class ApkInfo(object):
	packageName = None
	versionCode = 0
	versionName = None

	appName = None
	icLauncher = None

	manifestMetas = {}

	#构造方法
	def __init__(self,apk_path=None):
		if apk_path :
			infosList = os.popen("aapt d badging "+apk_path).readlines()
			self.loadInfo(infosList)

			self.loadMetas(os.popen("aapt dump xmltree "+apk_path+" AndroidManifest.xml").readlines())

	#获取应用基本信息
	def loadInfo(self,infolist):
		global packageName
		for info in infolist:
			key_dict = analyze_key_value(info)
			if 'package' in info and 'versionCode=' in info:
				self.packageName = key_dict['name']
				self.versionCode = key_dict['versionCode']
				self.versionName = key_dict['versionName']
				continue

			if 'application' in info and 'label=' in info and 'icon=' in info :
				self.appName = key_dict['label']
				self.icLauncher = key_dict['icon']

	# 获取应用meta信息
	def loadMetas(self,xml_tree):
		self.manifestMetas={}
		for i in range(0,len(xml_tree)):
			info = xml_tree[i].replace('\n','')
			if 'E: meta-data' in info:
				name = xml_tree[i+1].replace('\n','').split("=\"")[1].split("\"")[0]
				value =  (xml_tree[i+2].replace('\n','').split('type 0x')[1].split(')')[1]) if 'type 0x' in xml_tree[i+2] else xml_tree[i+2].replace('\n','').split("=\"")[1].split("\"")[0]
				self.manifestMetas[name]=value

File: src/model/apk_info.rb
class ApkInfo
  attr_accessor :package_name
  attr_accessor :version_code
  attr_accessor :version_name
  attr_accessor :app_name
  attr_accessor :ic_launcher
  attr_accessor :metas

  def initialize(apk_path)
    load_infos(`aapt d badging #{apk_path}`.split("\n"))

    load_metas(`aapt dump xmltree #{apk_path} AndroidManifest.xml`.split("\n"))
  end

  def to_string
    puts <<-EOB
appName\t#{self.app_name}
icLauncher\t#{self.ic_launcher}
packageName\t#{self.package_name}
versionCode\t#{self.version_code}
versionName\t#{self.version_name}
#{self.metas.inject([]){|acc,(k,v)| acc<<"meta:#{k}\t#{v}";acc}.join("\n")}
    EOB
  end

  private

  def load_infos(infos)
    infos.each do |info|
      key_dict = analyze_key_value(info)
      if info.include?('package') && info.include?('versionCode=')
        self.package_name = key_dict['name']
        self.version_code = key_dict['versionCode']
        self.version_name = key_dict['versionName']
        next
      end

      if info.include?('application') && info.include?('label=') && info.include?('icon=')
        self.app_name= key_dict['label']
        self.ic_launcher = key_dict['icon']
      end
    end
  end

  # 获取应用meta信息
  def load_metas(xml_tree)
    self.metas = {}
    xml_tree.each_with_index { |info, i|
      if info.include?('E: meta-data')
        name = xml_tree[i + 1].gsub("\n", '').split("=\"")[1].split("\"")[0]
        if xml_tree[i + 2].include?('type 0x')
          value = xml_tree[i + 2].gsub('\n', '').split('type 0x')[1].split(')')[1]
        elsif xml_tree[i + 2].include?('android:resource')
          value = xml_tree[i + 2].gsub("\n", '').split(')=')[1]
        elsif xml_tree[i + 2].include?('=@0x')
          value = "@" + xml_tree[i + 2].gsub("\n", '').split("=@")[1]
        else
          value = xml_tree[i + 2].gsub("\n", '').split("=\"")[1].split("\"")[0]
        end
        self.metas[name] = value
      end

    }
  end

  def analyze_key_value(xml_str='')
    key_dict = {}
    xml_str.split(" ").each do |package_item|
      item_array = package_item.split("=")
      if item_array.size != 2
        next
      end
      key_dict[item_array[0]] = item_array[1].gsub("'", '').gsub("\n", '')
    end
    key_dict
  end
end